### PR TITLE
implement manual encode/decode

### LINKED ISSUE #48: Custom implementation of Encode Decode for Leverage
Implement Encode / Decode for Leverage and encode it into u8 instead of u16. i.e. only store number of bit shifts. Use `trailing_zeros` to do the bit counting.

### PR DESCRIPTION
**Pull request by ermalkaleci**
Closes #48

### CODE REVIEW

File: modules/primitives/src/lib.rs [hunk 1]
```diff
@@ -78,6 +78,74 @@ bitmask! {
...
+	}
+}
+
+impl Encode for Leverage {
+	fn size_hint(&self) -> usize {
+		1
+	}
+
+	fn encode(&self) -> Vec<u8> {
+		let mut r: Vec<u8> = Vec::with_capacity(self.size_hint());
```
Comment (xlc) [inline modules/primitives/src/lib.rs]: `vec![u16::trailing_zeros(**self) as u8]`

File: modules/primitives/src/lib.rs [hunk 2]
```diff
@@ -78,6 +78,74 @@ bitmask! {
...
+		r.push(u16::trailing_zeros(**self) as u8);
+		r
+	}
+}
+
+impl Decode for Leverage {
+	fn decode<I: Input>(value: &mut I) -> Result<Self, Error> {
+		let trailing_zeros = value.read_byte()? as u16;
```
Comment (xlc) [inline modules/primitives/src/lib.rs]: no need to `as u16`

File: modules/primitives/src/lib.rs [hunk 3]
```diff
@@ -185,4 +253,27 @@ mod tests {
 		assert_eq!(Leverage::ShortThirty.value(), 30);
 		assert_eq!(Leverage::ShortFifty.value(), 50);
 	}
+
+	#[test]
+	fn encode_decode_should_work() {
+		let mut merged = LONGS.clone().to_vec();
+		merged.append(&mut SHORTS.clone().to_vec());
```
Comment (xlc) [inline modules/primitives/src/lib.rs]: use `extend_from_slice`

File: modules/primitives/src/lib.rs (hunk 2 above)
Comment (ermalkaleci) [reply modules/primitives/src/lib.rs]: @xlc forgot to cleanup

**Pull request merged**